shotover-proxy/tests/redis_int_tests/mod.rs

```diff
@@ -321,3 +322,54 @@ async fn cluster_dr() {
 
     shotover.shutdown_and_then_consume_events(&[]).await;
 }
+
+#[tokio::test(flavor = "multi_thread")]
+async fn test_failed_requests_metric_sink_single() {
+    let _compose = docker_compose("tests/test-configs/redis/passthrough/docker-compose.yaml");
+    let shotover = shotover_process("tests/test-configs/redis/passthrough/topology.yaml")
+        .start()
+        .await;
+    let mut connection = redis_connection::new_async("127.0.0.1", 6379).await;
+
+    redis::cmd("INVALID_COMMAND")
+        .arg("foo")
+        .query_async::<_, ()>(&mut connection)
+        .await
+        .unwrap_err();
+
+    // Redis client driver initialization sends 2 CLIENT SETINFO commands which trigger 2 errors
+    // because those commands are not available in the currently used redis version.
+    assert_metrics_key_value(
```

Comment:
I think we should move these assertions into assertion methods and then call them from redis_passthrough and redis_cluster_handling.
e.g. define a method named `assert_failed_requests_metric_is_incremented_on_error_response` that contains lines 329 to 348. We should include creation of shotover in the assertion since we need a fresh shotover to be able to make assertions on the state of the counter.
Alternatively we could avoid creating shotover in the assertion method and create a 2nd shotover in redis_passthrough/redis_cluster_handling.
I can see arguments for both approaches so I dont really mind which one we take.

The main reason is to reduce the time it takes to run the test suite.
On my machine, it takes 1 second to spin up `redis/passthrough/docker-compose.yaml` and 5 seconds to spin up `redis/cluster-handling/docker-compose.yaml`.
So by combining these cases into `redis_passthrough` and `redis_cluster_handling` we cut off 6 seconds from the test suite run time.
In isolation 6 seconds isnt much, but if every PR added 6 seconds we would have a problem.

The implementation within shotover looks good btw!

Reply:
Cool will do

Reply:
Just moved the test code into a new function `assert_failed_requests_metric_is_incremented_on_error_response()` to avoid spinning up a new redis container